src/ripple/app/misc/impl/AmendmentTable.cpp

```diff
@@ -554,40 +587,88 @@ AmendmentTableImpl::needValidatedLedger (LedgerIndex ledgerSeq)
 }
 
 void
-AmendmentTableImpl::doValidatedLedger(
+AmendmentTableImpl::doValidatedLedger (
     LedgerIndex ledgerSeq,
     std::set<uint256> const& enabled,
-    majorityAmendments_t const& majority)
+    majorityAmendments_t const& majority,
+    Rules const& rules)
 {
     for (auto& e : enabled)
         enable(e);
 
-    std::lock_guard sl(mutex_);
-    // Since we have the whole list in `majority`, reset the time flag, even if
-    // it's currently set. If it's not set when the loop is done, then any
-    // prior unknown amendments have lost majority.
-    firstUnsupportedExpected_.reset();
-    for (auto const& [ hash, time ] : majority)
     {
-        auto s = add(hash);
+        std::lock_guard sl(mutex_);
+        // Since we have the whole list in `majority`, reset the time flag,
+        // even if it's currently set. If it's not set when the loop is done,
+        // then any prior unknown amendments have lost majority.
+        firstUnsupportedExpected_.reset();
+        for (auto const& [ hash, time ] : majority)
+        {
+            auto s = add(hash, sl);
+
+            if (s->enabled)
+                continue;
+
+            if (!s->supported)
+            {
+                JLOG(j_.info()) << "Unsupported amendment " << hash
+                                 << " reached majority at " << to_string(time);
+                if (!firstUnsupportedExpected_ ||
+                    firstUnsupportedExpected_ > time)
+                {
+                        firstUnsupportedExpected_ = time;
+                }
+            }
+        }
+        if (firstUnsupportedExpected_)
+            firstUnsupportedExpected_ =
+                *firstUnsupportedExpected_ + majorityTime_;
+    }
+
+    if (rules.enabled(featureRetire2017Amendments))
+    {
+        // Remove retired amendments from the AmendmentMap_ so we stop
+        // reporting about amendments that are retired.  This should only
+        // need to be done once, so check whether it needs to be done
+        // before actually doing the work.
+        bool const needsRetiring = [this, ledgerSeq]()
+        {
+            std::lock_guard sl (mutex_);
+            lastUpdateSeq_ = ledgerSeq;
```

Comment:
Not in Scott's new code, but in existing code, I don't see lastUpdateSeq_ get updated. It is initialized to 0 and seems never changed. If so, the function `bool needValidatedLedger (LedgerIndex seq)` always return true, right? Again, it is in the existing code.

Reply:
Yes, this fixes a minor bug in the existing code.  I found the bug while testing the new code.  `needValidatedLedger()` always returning true was a pessimization of this code:  https://github.com/ripple/rippled/blob/develop/src/ripple/app/misc/AmendmentTable.h#L80  It would always call `doValidatedLedger()`.

Reply:
Right. And it is minor performance bug. I just did not get why the fix will only take effect after featureRetire2017Amendments is enabled.

Reply:
I don't remember the original motivation.  Perhaps I was being conservative.  Since I have an amendment readily available, putting it on the amendment means that the behavior change happens across the network at the same time.  I don't _think_ the change should impact behavior noticeably but, just in case, it's a bit more cautious to make the change on an amendment.

I'd be happy to move the change outside of the amendment if you think that's better.  I don't believe there's any risk to that.  But we would want to make the change carefully.  `lastUpdateSeq_` should only be touched under the lock.  And, for efficiency, we probably don't want to do the work inside the preceding `for` loop under its lock.  Where it is right now it a place where it's under a lock and it will only be executed once per call.

Reply:
Thanks for the explanation. I think it is fine as is. It is a minor performance bug and it should not be very long to get the amendment. I was just curious about your thinking.